### ISSUE TITLE
"Failed to open file for reading"

### ISSUE THREAD

**Issue by HubSteven** [lhm0/rotating_display_color#2]
Hi,
I've flashed both the ESP32 and the Raspberry Pi Pico on the pcb.
The micro SD card also has the necessary files in the root directory.
However the LED's won't light up and I don't get any webpage when I'm connected to the SSID.
The ESP32 show's the following in the terminal connection over  USB:

> start initialization....
SD Card Type: SDSC
SD Card Size: 121MB
nun folgt .read_f()
Failed to open file for reading
.read_f() abgeschlossen
Failed to open file for reading
Access Point started.
IP-Adresse: 192.168.4.1
Failed to open file for reading
SD Card Type: SDSC
SD Card Size: 121MB
der Server wurde gestartet
Zeit: 1:0:2
request weather data.....
Wifi connection lost...
ERROR: 
deserializeJson() failed: EmptyInput
icon: 0
temp: 0.0 C
humidity: null %
Failed to open file for reading
Failed to open file for reading
Failed to open file for reading
Failed to open file for reading
Failed to open file for reading
Failed to open file for reading
Failed to open file for reading
Failed to open file for reading
Failed to open file for reading
Failed to open file for reading
Failed to open file for reading

I've tried several microSD cards with different sizes.

**Comment by lhm0** [comment]
Hi Hub,

from the serial output we can read, that

1. the ESP32 microcontroller is running. That is good news.
2. there should be a Wifi access point ("RD56c"). If you can connect to this AP (even without any response from the server) is another good sign. It means, that the WiFi function is working as well.
3. the microcontroller can talk to the SD card. Otherwise, the output would be "card mount failed". But you get "SD Card Type: SDSC, SD Card Size: 121MB". This means, card mount was successful. Good news.

Given 3, it sounds like the software does not find the files it is looking for. Let's check the file structure:

the folders "gif", "html", "images", and "variables" all need to be in the root directory. For the initialization the "variables" folder needs to have the various *.txt" files, as you can see from the "data" folder of the repository.

It really looks like the ESP32 is very close to full function.

I am more concerned about the RP2040, as you do not see any flickering of the LEDs. This seems to indicate, that the RP2040 is not running. Did the software upload work for the RP2040? Are there any messages on the serial port?

Best regards,

Ludwin

**Comment by lhm0** [comment]
another thought if the file structure of the SD card is correct: what format does the SD card have? You need to use FAT32.

I had some trouble in the past with formatting the SD card under Windows. If problems persist, you might want to re-format the card with https://www.sdcard.org/downloads/formatter/ . This program works for sure.

**Comment by lhm0** [comment]
I check on the RP2040: there should be a message "brightness: 0" once every second on the USB port. 

If you use the Visual Studio Code terminal for monitoring, it might be tricky to connect to the microcontroller.

**Comment by HubSteven** [comment]
Thanks for your quick reply. I do indeed have the "brightness: 0"message from the RP2040.
My directory structure on the SD card look like this:
![image](https://github.com/user-attachments/assets/c6897a4f-54a8-4ffa-8955-f0bef13fd250)

**Comment by HubSteven** [comment]
Also tried the sd card formatter tool, but still the same thing.
I don't have the DC motor yet so the rd56c pcb is on top of the PS  PCB with a piece of paper in between to prevent shorts.
Without the PCB rotating and without SD card, the LED's should still light up for a test pattern?

**Comment by lhm0** [comment]
yes, even without rotation there should be a flickering. Also, it does not matter, what the ESP32 is doing. The SD card is irrelevant for the flickering.

Let me look into the SD card issue. Very, very strange

**Comment by HubSteven** [comment]
My 5V supply is dead, got some hardware fixing to do. I'll let you know once it works.
BTW I've noticed that on the silkscreen C310 and C311 are swapped.
Another thing to take into account is that the DC motor apparently has two versions. I've orderd the wrong one. 
![image](https://github.com/user-attachments/assets/922d65fc-f426-4639-96e6-4c0af27ed3ad)
This one has the same type number but a different piece of plastic with 3 screw holes instead of four.

**Comment by lhm0** [comment]
ok. That explains, why you do not see any flickering.

regarding the motor, there are several versions, unfortunately. The one that you have might still work, as long as the height of the assembly is the same.

**Comment by lhm0** [comment]
I have added a test program for the SD card reader. Please check the subfolder "RD56c_ESP32_SD_Test" of the repository and upload to the ESP32.

The software should list all files of the SD card via the Serial monitor.

**Comment by HubSteven** [comment]
Just fixed the sort on the 5V line, the LEDs light up now (red and blue flickering).
From your sd test code I get this:
![image](https://github.com/user-attachments/assets/9c47f87e-2ac8-4950-9ab4-363eb861f8b6)

**Comment by lhm0** [comment]
there were a couple of inconsistencies, which I have corrected. Please download the latest repository and upload the code to the ESP32. I hope things will work, now.

Great to hear, that you have fixed the 5V supply. There should be a "red-green-blue" sequence with different phases on the left and the right half of the LED row.

If the ESP32 works correctly, the "red-green-blue" sequence should be replaced with some white flickering after 10 seconds, or so. 

Try out the web interface. It should work, now.

Keep me posted. I will check tomorrow.

**Comment by HubSteven** [comment]
Great, it finally works! I'll keep this issue opened to post you a picture of the whole assembly once I've got the DC motor in.

**Comment by lhm0** [comment]
Dear Hub, I spent an afternoon on bug fixing the software. The file browser is much more stable now, though still not 100% perfect. But it is usable. Furthermore, I have added a new start screen background with a "real gear". 
As you might know, I have been working on a single microcontroller display (monochrome) for a while. However, in comparison, the image stability of the color display is so much better. There is no jitter, no instability. The dual controller approach might have its disadvantages, but the performance is simply a different class.

**Comment by HubSteven** [comment]
Thank you, got it uploaded to the ESP32. I'll let you know once I got the motor in.

**Comment by HubSteven** [comment]
I got the cd motor in. I was only wondering if the slit of the pcb with the two m2 nuts should be inserted over the indicated piece of plastic. Because it seems to be too thick to fit in the slit of the pcb.

![Image](https://github.com/user-attachments/assets/45eb6744-b243-4806-b327-b1f1cfee08c3)

**Comment by lhm0** [comment]
here is how it looks with my motor:

![Image](https://github.com/user-attachments/assets/6e3b0319-3af2-4d32-975b-73fe8414b62f)

![Image](https://github.com/user-attachments/assets/a41a6695-a0f7-4ae3-b21a-024e0c19564b)

**Comment by galmiklos** [comment]
Hi Steven,
Sorry for contacting you here, I couldn't find a way to private message you. I am thinking about building it, too, and was trying to find PCBs, so I reached out to Ludwin in the comments section on youtube, and he pointed out that you are in the middle of building, and that you may have an extra set available. I know when ordering, the minimum number is usually 5, so before I spend $80 to order 5 sets, I was hoping to find someone with surplus boards. If you can spare a set or two, it could also lower your cost overall, I'm willing to "chip in".From your profile and the discussion I couldn't tell where you were located, hopefully on this side of the pond, so that shipping would not be too expensive. I live in Texas. You can email me directly at miklos.gal@gmail.com, too.
Thanks,
MIki.

**Comment by HubSteven** [comment]
@galmiklos I'm located in the Netherlands, so I don't think sending you my PCBs will make it any cheaper for you. Besides I've hand soldered every SMD component, that's something I would not recommend. Getting them pre-assembled is much easier.

@lhm0 I've put together the whole assembly and it fully works, only got some problems with balancing so it sound a bit like an old Diesel right now haha.

Only problem I have is with changing the animation logo's. Is this what the webpage should look like?

![Image](https://github.com/user-attachments/assets/42ce4dd9-ee7e-4dc1-8872-51d1e056a6df)

**Comment by lhm0** [comment]
now I am really impressed: you hand soldered this PCB? You must be an artist.

Yes, this is the right screen for the data conversion. You can convert any image (including GIFs). 

1. Tap on "load image". A file selection box will come up, which allows you to select and load an image from your mobile phone to the display. 
2. The upper file selection box should show the contents of your SD card. I am not sure why it does not do so, in your screen shot. Maneuver to a folder, where you would like to store the converted image.
3. you can either scale or crop the image. Tap "save .rdc". Now, the controller will save the image to the SD card. Please note: it will take quite a while until this process is completed. Depending on the image size it can easily take a minute or more.
4. now you need to assign the file to either the clock or the image mode. long-tap on the filename in the upper selection box, then tap on "select logo" or "select image".

**Comment by galmiklos** [comment]
@HubSteven , yes, that's what I do too, hand solder the components. I usually make my own PCBs, including solder mask and silk screen, but I only can do double sided. But yes, shipping from the Netherlands is not an option. Thanks for taking the time to answer.
